File: ladim/ibms/surface_light.f90
! compile with gfortran -fdec-math

program test

  integer, dimension(6) :: timevec
  real :: lat
  real :: slig

timevec = (/ 2014, 6, 23, 12, 0, 0 /)
lat = 23.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig

  
timevec = (/ 2014, 6, 23, 12, 0, 0 /)
lat = 60.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig

timevec = (/ 2014, 6, 23, 18, 0, 0 /)
lat = 60.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig

timevec = (/ 2014, 6, 23, 18, 0, 0 /)
lat = 80.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig

timevec = (/ 2014, 6, 23, 0, 0, 0 /)
lat = 60.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig

timevec = (/ 2014, 6, 23, 0, 0, 0 /)
lat = 80.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig



timevec = (/ 2014, 12, 24, 12, 0, 0 /)
lat = 80.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig

timevec = (/ 2014, 12, 24, 12, 0, 0 /)
lat = 60.0
call surf_light(timevec, lat, slig)
print *,  timevec, lat, slig





end program test



!     -------------------------------------------------
      SUBROUTINE surf_light(timevector,B,SLIG)
!     -------------------------------------------------
!   Surface irradiance after Skartveit & Olseth 1988
      implicit none
      integer, dimension(6) :: timevector !y m d h min s
      integer :: i
      integer, dimension(12) :: days_in_month
      REAL     B,D,DELTA,HEIGHT,H,P,V,MAXLIG
      REAL     H12,TWLIGHT
      real :: deg
      real :: SLIG


!     B:Degrees north
!     DELTA: sun declination
!     D: day of the year
!     H: hour of day
!     HEIGHT: sin(sunheight)
!     IRR. irradiance above sea surface uEm-2s-1
!     P: Pi
!     R: factor for distance variations between sun-earth
!     SLIG:surface light
!     V: sunheight in degrees
!     TWLIGHT: light at 0-degree sun
!   MAXLIG: level of irradiance at midday

!FV
      MAXLIG = 1500 ! varierer mellom <200,2000>?!
      D = 0
      days_in_month(1:12) =                                             &
         (/31,28,31,30,31,30,31,31,30,31,30,31/)
      do i = 1,timevector(2)
      D = D + days_in_month(i)
      end do
      D = D - days_in_month(timevector(2)) + timevector(3)
      H = timevector(4)
!FV


      P = 3.1415927
      deg = 180.0 / P
      TWLIGHT = 5.76
        DELTA = .3979*SIN((.9856*(D-80)+ 1.9171*(SIN(.9856*D/deg)-.98112))/deg)
        H12 = DELTA*SIN(B/deg)- SQRT(1.-DELTA**2)*COS(B/deg)*COS(15.*12/deg)
        HEIGHT = DELTA*SIN(B/deg)- SQRT(1.-DELTA**2)*COS(B/deg)*COS(15.*H/deg)

        V = deg * ASIN(HEIGHT)

          IF (V .GE. 0.) THEN
            SLIG = MAXLIG*(HEIGHT/H12) + TWLIGHT
          ELSE IF (V .GE. -6.) THEN
            SLIG = ((TWLIGHT - .048)/6.)*(6.+V)+.048
          ELSE IF (V .GE. -12.) THEN
            SLIG = ((.048 - 1.15E-4)/6.)*(12.+V)+1.15E-4
          ELSE IF (V .GE. -18) THEN
            SLIG = (((1.15E-4)-1.15E-5)/6.)*(18.+V)+1.15E-5
          ELSE
            SLIG = 1.15E-5
          ENDIF


  end subroutine surf_light

File: ladim/ibms/surface_light.py
# Surface ligth module
# After Skartveit & Olseth 1988
import datetime
import numpy as np

pi = np.pi
sin = np.sin
cos = np.cos
rad = np.pi / 180.0
deg = 180/pi


def sind(x):
    return np.sin(x*rad)


def cosd(x):
    return np.cos(x*rad)


def asind(v):
    return deg*np.arcsin(v)


def surface_light(dtime, lat):

    # def surface_light(timevector, B):
    # B = latitude (north)
    # B:Degrees north
    #  DELTA: sun declination
    #  D: day of the year
    #  H: hour of day
    #  HEIGHT: sin(sunheight)
    #  IRR. irradiance above sea surface uEm-2s-1
    #  P: Pi
    #  R: factor for distance variations between sun-earth
    #  SLIG:surface light
    #  V: sunheight in degrees
    #  TWLIGHT: light at 0-degree sun
    # MAXLIG: level of irradiance at midday

    maxlight = 1500  # value between 200 and 2000

    time_tuple = dtime.timetuple()
    # day of year, original does not consider leap years
    yday = time_tuple.tm_yday
    # hours in local time ot UTC???
    hours = time_tuple.tm_hour

    twilight = 5.76
    delta = .3979 * sind(.9856*(yday-80) + 1.9171*(sind(.9856*yday)-.98112))
    h12 = delta*sind(lat) - np.sqrt(1.-delta**2)*cosd(lat)*cosd(15.*12)
    height = delta*sind(lat) - np.sqrt(1.-delta**2)*cosd(lat)*cosd(15.*hours)

    # print(delta, h12)
    v = asind(height)  # sun height [deg]
    # print(height, v)

    if (v >= 0.):
        slig = maxlight*(height/h12) + twilight
    elif (v >= -6.):
        slig = ((twilight - .048)/6.)*(6.+v)+.048
    elif (v >= -12.):
        slig = ((.048 - 1.15e-4)/6.)*(12.+v)+1.15e-4
    elif (v >= -18):
        slig = (((1.15e-4)-1.15e-5)/6.)*(18.+v)+1.15e-5
    else:
        slig = 1.15e-5

    return slig

if __name__ == '__main__':

    dtime = datetime.datetime(2014, 6, 23, 12, 0, 0)
    lat = 60.0
    print(surface_light(dtime, lat))

    dtime = datetime.datetime(2014, 6, 23, 18, 0, 0)
    lat = 60.0
    print(surface_light(dtime, lat))
